cdc/sink/pulsar/option.go

```
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+package pulsar
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+)
+
+// Option represents the options needed by the pulsar client
+type Option struct {
+	clientOptions   *pulsar.ClientOptions
+	producerOptions *pulsar.ProducerOptions
+}
+
+const route = "$route"
+
+func parseSinkOptions(u *url.URL) (opt *Option, err error) {
+	switch u.Scheme {
+	case "pulsar", "pulsar+ssl":
+	default:
+		return nil, fmt.Errorf("unsupported pulsar scheme: %s", u.Scheme)
+	}
+	c, err := parseClientOption(u)
+	if err != nil {
+		return nil, err
+	}
+	p, err := parseProducerOptions(u)
+	if err != nil {
+		return nil, err
+	}
+	opt = &Option{
+		clientOptions:   c,
+		producerOptions: p,
+	}
+
+	p.MessageRouter = func(message *pulsar.ProducerMessage, metadata pulsar.TopicMetadata) int {
+		partition, _ := strconv.Atoi(message.Properties[route])
+		message.Properties = nil
+		return partition
+	}
+	return
+}
+
+func parseClientOption(u *url.URL) (opt *pulsar.ClientOptions, err error) {
+	vs := values(u.Query())
+	opt = &pulsar.ClientOptions{
+		URL:                        (&url.URL{Scheme: u.Scheme, Host: u.Host}).String(),
+		ConnectionTimeout:          vs.Duration("connectionTimeout"),
+		OperationTimeout:           vs.Duration("operationTimeout"),
+		TLSTrustCertsFilePath:      vs.Str("tlsTrustCertsFilePath"),
+		TLSAllowInsecureConnection: vs.Bool("tlsAllowInsecureConnection"),
+		TLSValidateHostname:        vs.Bool("tlsValidateHostname"),
+		MaxConnectionsPerBroker:    vs.Int("maxConnectionsPerBroker"),
+	}
+	auth := vs.Str("auth")
+	if auth == "" {
+		if u.User.Username() == "" {
+			// no auth
+			return opt, nil
+		}
+		// use token provider by default
+		opt.Authentication = pulsar.NewAuthenticationToken(u.User.Username())
+		return opt, nil
+	}
+	param := jsonStr(vs.SubPathKV("auth"))
+	opt.Authentication, err = pulsar.NewAuthentication(auth, param)
```

this error is ignored by mistake